maintainers_annotation/maintainers_annotation.go

```diff
@@ -0,0 +1,190 @@
+}
+
+// Annotation groups the annotation reason and, in the future, the related probe.
+// If there is a probe, the reason applies to the probe.
+// If there is not a probe, the reason applies to the check or group of checks in
+// the exemption.
+type Annotation struct {
+	AnnotationReason AnnotationReason `yaml:"annotation"`
+}
+
+// Exemption defines a group of checks that are being annotated for various reasons.
+type Exemption struct {
+	Checks      []string     `yaml:"checks"`
+	Annotations []Annotation `yaml:"annotations"`
+}
+
+// MaintainersAnnotation is a group of exemptions.
+type MaintainersAnnotation struct {
+	Exemptions []Exemption `yaml:"exemptions"`
+}
+
+// parseScorecardYmlFile takes the scorecard.yml file content and returns a `MaintainersAnnotation`.
+func parseScorecardYmlFile(ma *MaintainersAnnotation, content []byte) error {
+	unmarshalErr := yaml.Unmarshal(content, ma)
+	if unmarshalErr != nil {
+		return sce.WithMessage(sce.ErrScorecardInternal, unmarshalErr.Error())
+	}
+
+	return nil
+}
+
+// validateAndParseScorecardYmlFile takes the scorecard.yml files found, validates if its the correct
+// configuration file and parse its configurations.
+func validateAndParseScorecardYmlFile(path string, content []byte, args ...interface{}) (bool, error) {
+	// Verify if received the necessary parameters to save the scorecard YML configurations
+	if len(args) != 1 {
+		return false, fmt.Errorf(
+			"validateAndParseScorecardYmlFile requires exactly one argument to save maintainers annotation data")
+	}
+	ma, ok := args[0].(*MaintainersAnnotation)
+	if !ok {
+		return false, fmt.Errorf(
+			"validateAndParseScorecardYmlFile requires argument of type *MaintainersAnnotation")
+	}
+
+	// Exclude scorecard GitHub Action file
+	if strings.Contains(path, ".github") {
+		return true, nil
+	}
+
+	// Keep looking for files if content is empty
+	if len(content) == 0 {
+		return true, nil
+	}
+
+	// Found scorecard.yml file, parse file
+	err := parseScorecardYmlFile(ma, content)
+	if err != nil {
+		return false, err
+	}
+
+	return false, nil
+}
+
+// readScorecardYmlFromRepo reads the scorecard.yml file from the repo and returns its configurations.
+func readScorecardYmlFromRepo(repoClient clients.RepoClient) (MaintainersAnnotation, error) {
+	ma := MaintainersAnnotation{}
+	err := fileparser.OnMatchingFileContentDo(repoClient, fileparser.PathMatcher{
+		Pattern:       "scorecard.yml",
+		CaseSensitive: false,
+	}, validateAndParseScorecardYmlFile, &ma)
+
+	if err != nil {
+		return ma, fmt.Errorf("%w", err)
+	}
+
+	if ma.Exemptions == nil {
+		return ma, errNoScorecardYmlFile
+	}
+
+	return ma, nil
+}
+
+// GetMaintainersAnnotation reads the maintainers annotation from the repo, stored in scorecard.yml, and returns it.
+func GetMaintainersAnnotation(repoClient clients.RepoClient) (MaintainersAnnotation, error) {
```

Comment:
what do you think about:
```go
func Parse(r io.Reader) (MaintainersAnnotation, error)
```

The annotations don't know need to know about the repo clients, just parse what they're given.

Reply:
Trying to address this change, I've got:
```diff
# scorecard.go

	// Get configuration
-	c, err := config.Parse(repoClient)
+	content, err := repoClient.GetFileContent("scorecard.yml")
+	// If configuration file exists, continue. Otherwise, ignore
+	if err == nil {
+		c, err := config.Parse(bytes.NewReader(content))
		if err != nil {
			return ScorecardResult{}, err
		}
		ret.Config = c
+	}

```

```diff
# config.go

// Parse reads the configuration file from the repo, stored in scorecard.yml, and returns a `Config`.
- func Parse(repoClient clients.RepoClient) (Config, error) {
+ func Parse(r io.Reader) (Config, error) {
	c := Config{}
-	// Find scorecard.yml file in the repository's root
-	content, err := repoClient.GetFileContent("scorecard.yml")
+	content, err := io.ReadAll(r)
	if err != nil {
-		if errors.Is(err, os.ErrNotExist) {
-			// If scorecard.yml doesn't exist, ignore
-			return c, nil
-		}
		return c, fmt.Errorf("fail to read configuration file: %w", err)
	}

	err = parseFile(&c, content)
	if err != nil {
		return c, fmt.Errorf("fail to parse configuration file: %w", err)
	}

	// Return configuration
	return c, nil
}

```

Is this what you were expecting?

Reply:
Ack, thanks! And thanks for reminding about closing the file.

Reply:
Done.